File: flink-java-code/src/main/java/com/atguigu/day2/SensorSource.java
package com.atguigu.day2;

import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;

import java.util.Calendar;
import java.util.Random;

public class SensorSource extends RichParallelSourceFunction<SensorReading> {
    private boolean running = true;

    @Override
    public void run(SourceContext<SensorReading> ctx) throws Exception {
        Random rand = new Random();

        String[] sensorIds = new String[10];
        double[] curFTemp = new double[10];

        for (int i = 0; i < 10; i++) {
            sensorIds[i] = "sensor_" + i;
            curFTemp[i] = 65 + (rand.nextGaussian() * 20);
        }

        while (running) {
            long curTime = Calendar.getInstance().getTimeInMillis();
            for (int i = 0; i < 10; i++) {
                curFTemp[i] = rand.nextGaussian() * 0.5;
                ctx.collect(new SensorReading(sensorIds[i], curTime, curFTemp[i]));
            }

            Thread.sleep(1000L);
        }


    }

    @Override
    public void cancel() {
        running = false;
    }
}

File: flink-scala-code/src/main/scala/com/atguigu/day2/SensorSource.scala
package com.atguigu.day2

import java.util.Calendar

import org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext
import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction

import scala.util.Random

class SensorSource extends RichParallelSourceFunction[SensorReading] {
  var running = true

  override def run(ctx: SourceContext[SensorReading]): Unit = {
    val rand = new Random

    var curFTemp = (1 to 10).map(i => ("sensor_" + i, rand.nextGaussian() * 20))

    // 无限发送数据
    while (running) {
      curFTemp = curFTemp.map(t => (t._1, t._2 + (rand.nextGaussian() * 0.5)))

      // 产生一个毫秒时间戳
      val curTime = Calendar.getInstance.getTimeInMillis

      curFTemp.foreach(t => ctx.collect(SensorReading(t._1, curTime, t._2)))

      Thread.sleep(300)
    }
  }

  // 取消发送数据
  override def cancel(): Unit = running = false
}